h/browser/chrome/lib/uri-info.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * A wrapper for the server's uriinfo endpoint.
+ *
+ * Exports a function that makes an HTTP request to the server's uriinfo
+ * endpoint and returns a Promise that resolves to the response. Usage:
+ *
+ *     var uriInfo = require('uri-info');
+ *     uriInfo(uri).then(function(info) {
+ *       ...
+ *     });
+ *
+ * This function can be called with the same uri multiple times consecutively
+ * and it'll just keep returning the same Promise - won't make multiple HTTP
+ * requests.
+ *
+ * But if you call the function with a different uri, then call it with the
+ * first uri again, then it _will_ make another request for the first uri.
+ *
+ * Returns a rejected Promise if the HTTP request times out, fails or is
+ * invalid for any reason.
+ *
+ * Returns a rejected Promise is the given uri is undefined.
+ *
+ * @param {String} uri The URI to be checked
+ *
+ * @returns {Promise} A Promise that resolves to an object with two properties:
+ *   1. total:   the number of annotations there are of this URI (Number)
+ *   2. blocked: whether or not this URI is blocklisted (Boolean)
+ *
+ *
+ * @module
+ */
+'use strict';
+
+var settings = require('./settings');
+var uriInfoPromise;
+
+/**
+  * Make an HTTP request to the uriinfo endpoint and return a Promise that
+  * resolves to the value of the response.
+  *
+  */
+function getUriInfo(uri) {
+  if (uri === undefined) {
+    return Promise.reject({reason: 'uri was undefined'});
+  }
+
+  var p = new Promise(function(resolve, reject) {
+    var request = new XMLHttpRequest();
+    request.timeout = 1000;
+
+    request.onload = function() {
+      var info;
+      var total;
+      var blocked;
+      var errorMsg;
+
+      try {
+        info = JSON.parse(this.responseText);
+        total = info.total;
+        blocked = info.blocked;
+      } catch (e) {
+        errorMsg = 'Received invalid JSON from the server: ';
+        errorMsg = errorMsg + this.responseText;
+        console.error(errorMsg);
```

Comment:
I think I'd rather leave these here. If I move the error logging into a catch or onRejected in the code that's using the promise, then each error will get logged twice (once when the promise is called when the page loads, again when the same promise is called again when the user clicks the button to launch h), and if you refresh the page the error will also be logged again even though the http request wasn't made again. With the logging where it is, it's one failed http request = one logged error.